monai/utils/pixel_nms.py

```diff
@@ -0,0 +1,83 @@
+
+
+class PixelNMS:
+    """
+    Performs pixel based non-maximum suppression (NMS) on the probabilities map via
+    iteratively selecting the coordinate with highest probability and then move it as well
+    as its surrounding values. The remove range is determined by the parameter `pixel_dis`.
+    If multiple coordinates have the same highest probability, only one of them will be
+    selected.
+
+    Args:
+        spatial_dims: number of spatial dimensions of the input probabilities map.
+        sigma: the standard deviation for gaussian filter.
+            It could be a single value, or `spatial_dims` number of values. Defaults to 0.
+        prob_threshold: the probability threshold, the PixelNMS will stop searching if
+            the highest probability is no larger than the threshold. Defaults to 0.5.
+        pixel_dis: the distance in pixels that determines the surrounding area of the selected
+            coordinate that will be removed before the next iteration. The area is a square (cube)
+            for 2D (3D) input, and the edge length is 2 * pixel_dis. Defaults to 24.
+
+    Return:
+        a list of selected lists, where inner lists contain probability and coordinates.
+        For example, for 3D input, the inner lists are in the form of [probability, x, y, z].
+
+    """
+
+    def __init__(
+        self,
+        spatial_dims: int,
+        sigma: Union[Sequence[float], float, Sequence[torch.Tensor], torch.Tensor] = 0,
+        prob_threshold: float = 0.5,
+        pixel_dis: int = 24,
+    ) -> None:
+        self.sigma = sigma
+        self.spatial_dims = spatial_dims
+        if self.sigma != 0:
+            self.filter = GaussianFilter(spatial_dims=spatial_dims, sigma=sigma)
+        self.prob_threshold = prob_threshold
+        self.pixel_dis = pixel_dis
+
+    def __call__(
+        self,
+        probs_map: Union[np.ndarray, torch.Tensor],
+        resolution_level: int = 0,
+    ):
+        """
+        probs_map: the input probabilities map, it must have shape (H[, W, ...]).
+        resolution_level: the level at which the original input is made. The returned
+            coordinates will be converted according to this value.
+        """
+        if self.sigma != 0:
+            if not isinstance(probs_map, torch.Tensor):
+                probs_map = torch.as_tensor(probs_map, dtype=torch.float)
+            self.filter.to(probs_map)
+            probs_map = self.filter(probs_map)
+
+        if isinstance(probs_map, torch.Tensor):
+            probs_map = probs_map.detach().cpu().numpy()
+
+        probs_map_shape = probs_map.shape
+        resolution = pow(2, resolution_level)
+
+        outputs = []
+        while np.max(probs_map) > self.prob_threshold:
+            max_idx = np.unravel_index(probs_map.argmax(), probs_map_shape)
+            prob_max = probs_map[max_idx]
+            max_idx_arr = np.asarray(max_idx)
+            coord_wsi = ((max_idx_arr + 0.5) * resolution).astype(int)
```

Comment:
this `resolution` step seems to be pathology specific, shall we remove it from this generic utility? also the name of the method seems to suggest 2D (pixels but not voxels), perhaps it could be `ProbNMS` because it only works for probability scores

Reply:
Hi @behxyz , what do you think? Remove `resolution` should be fine and we can use some post processing method to do it right?

Reply:
Hi, removing the resolution requires the user to manually change the resolution before calling it and then change it back after calling it, which prone to error. Maybe we should put it as an option or so.

Reply:
I feel putting a resolution option wouldn't resolve any issue without having access to the full meta data of the input slide...we can add a resolution based wrapper in the `monai.apps.pathology` module, but this generic module should be decoupled from that concept. what do you think?

Reply:
@yiheng-wang-nv @wyli 
Although it doesn't require the access to meta data, it needs the knowledge of the resolution the inference has performed at. So I agree, we can make this one a generic `ProbNMS` without resolution and add another one for pathology in `monai.apps.pathology` to handle.